File: src/kotlin/standalone/sessionless/src/main/kotlin/com/planetnine/sessionless/impl/Sessionless.kt
package com.planetnine.sessionless.impl

import com.planetnine.sessionless.impl.Sessionless.WithCustomVault
import com.planetnine.sessionless.impl.Sessionless.WithKeyStore
import com.planetnine.sessionless.impl.exceptions.KeyPairNotFoundException
import com.planetnine.sessionless.models.ISessionless
import com.planetnine.sessionless.models.vaults.ICustomVault
import com.planetnine.sessionless.models.vaults.IVault
import com.planetnine.sessionless.util.KeyUtils
import com.planetnine.sessionless.util.KeyUtils.toECHex
import com.planetnine.sessionless.util.KeyUtils.toECPrivateKey
import com.planetnine.sessionless.util.KeyUtils.toECPrivateKeyParameters
import com.planetnine.sessionless.util.KeyUtils.toECPublicKeyParameters
import com.planetnine.sessionless.util.KeyUtils.toHex
import com.planetnine.sessionless.util.hashKeccak256
import org.bouncycastle.crypto.signers.ECDSASigner
import java.security.KeyPair
import java.security.PrivateKey
import java.security.interfaces.ECPrivateKey
import java.util.UUID

/** [Sessionless] implementation (sealed, check the subclasses)
 * @see WithKeyStore
 * @see WithCustomVault */
sealed class Sessionless(override val vault: IVault) : ISessionless {
    /** [ISessionless.WithKeyStore] implementation with [Sessionless] as superclass */
    class WithKeyStore(override val vault: KeyStoreVault) :
        Sessionless(vault),
        ISessionless.WithKeyStore {

        override fun generateKeys(keyAccessInfo: KeyAccessInfo): KeyPair {
            val pair = KeyUtils.generateKeyPair()
            vault.save(pair, keyAccessInfo)
            return pair
        }

        override suspend fun generateKeysAsync(keyAccessInfo: KeyAccessInfo): KeyPair {
            val pair = KeyUtils.generateKeyPairAsync()
            vault.save(pair, keyAccessInfo)
            return pair
        }

        override fun getKeys(keyAccessInfo: KeyAccessInfo): KeyPair =
            vault.get(keyAccessInfo)

        override fun sign(message: String, keyAccessInfo: KeyAccessInfo): MessageSignatureHex {
            val privateKey = getKeys(keyAccessInfo).private
            return sign(message, privateKey)
        }

        override fun verifySignature(
            signedMessage: SignedMessageWithECKey,
            keyAccessInfo: KeyAccessInfo
        ): Boolean {
            val publicKey = getKeys(keyAccessInfo).public
            val withKey = signedMessage.withKey(publicKey)
            return verifySignature(withKey)
        }
    }

    /** [ISessionless.WithCustomVault] implementation with [Sessionless] as superclass */
    class WithCustomVault(override val vault: ICustomVault) :
        Sessionless(vault),
        ISessionless.WithCustomVault {

        override fun generateKeys(): KeyPairHex {
            val pair = KeyUtils.generateKeyPair()
            val simple = pair.toECHex()
            vault.save(simple)
            return simple
        }

        override suspend fun generateKeysAsync(): KeyPairHex {
            val pair = KeyUtils.generateKeyPairAsync()
            val simple = pair.toECHex()
            vault.save(simple)
            return simple
        }

        override fun getKeys(): KeyPairHex? = vault.get()

        override fun sign(message: String): MessageSignatureHex {
            // throw if not found (caller should not call this before generating keys)
            val privateString = getKeys()?.privateKey
                ?: throw KeyPairNotFoundException()
            val privateKey = privateString.toECPrivateKey(KeyUtils.Defaults.parameterSpec)
            return sign(message, privateKey)
        }

        override fun verifySignature(signedMessage: SignedMessage): Boolean {
            val pair = getKeys() ?: throw KeyPairNotFoundException()
            val withKey = signedMessage.withKey(pair.publicKey)
            return verifySignature(withKey)
        }
    }

    override fun sign(message: String, privateKey: PrivateKey): MessageSignatureHex {
        val privateHex = (privateKey as ECPrivateKey).toHex()
        val privateKeyParameters = privateHex.toECPrivateKeyParameters()
        val signer = ECDSASigner().apply {
            init(true, privateKeyParameters)
        }
        val messageHash = message.hashKeccak256()
        val signatureArray = signer.generateSignature(messageHash)
        return MessageSignatureInt(signatureArray).toHex()
    }

    override fun verifySignature(signedMessage: SignedMessageWithKey): Boolean {
        val publicKeyParameters = signedMessage.publicKey.toECPublicKeyParameters()
        val withECKey = signedMessage.withKey(publicKeyParameters)
        return verifySignature(withECKey)
    }

    override fun verifySignature(signedMessage: SignedMessageWithECKey): Boolean {
        val signer = ECDSASigner().apply {
            init(false, signedMessage.publicKey)
        }
        val messageHash = signedMessage.message.hashKeccak256()
        val signatureInt = signedMessage.signature.toInt()
        return signer.verifySignature(
            messageHash,
            signatureInt.r,
            signatureInt.s
        )
    }

    override fun generateUUID(): String = UUID.randomUUID().toString()


    override fun associate(vararg signedMessages: SignedMessageWithKey): Boolean {
        signedMessages.twoOrThrow()
        return signedMessages.all(::verifySignature)
    }

    override fun associate(vararg signedMessages: SignedMessageWithECKey): Boolean {
        signedMessages.twoOrThrow()
        return signedMessages.all(::verifySignature)
    }

    companion object {
        private fun <T> Array<T>.twoOrThrow() {
            if (size < 2) {
                throw IllegalArgumentException("Must have at least two messages to associate")
            }
        }
    }
}

File: src/csharp/Sessionless/Impl/Sessionless.cs
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;
using SessionlessNET.Impl.Exceptions;
using SessionlessNET.Models;
using SessionlessNET.Util;

namespace SessionlessNET.Impl;

/// <summary> <see cref="ISessionless"/> implementation </summary>
/// <param name="vault"> The way to store and retrieve <see cref="KeyPairHex"/>s </param>
public class Sessionless(IVault vault) : ISessionless {
    public IVault Vault { get; } = vault;

    public string GenerateUUID() => Guid.NewGuid().ToString();
    public KeyPairHex GenerateKeys() {
        var pair = KeyUtils.GenerateKeyPair();
        return StoreHexPair(pair);
    }

    public async Task<KeyPairHex> GenerateKeysAsync() {
        var pair = await KeyUtils.GenerateKeyPairAsync();
        return StoreHexPair(pair);
    }

    private KeyPairHex StoreHexPair(AsymmetricCipherKeyPair pair) {
        var simple = pair.ToHex();
        Vault.Save(simple);
        return simple;
    }


    public KeyPairHex? GetKeys() => Vault.Get();


    public MessageSignatureHex Sign(string message) {
        var privateHex = GetKeys()?.PrivateKey
            ?? throw new KeyPairNotFoundException();
        return Sign(message, privateHex);
    }

    public MessageSignatureHex Sign(string message, string privateKeyHex) {
        if (!privateKeyHex.IsBytes()) {
            throw new HexFormatRequiredException(nameof(privateKeyHex));
        }
        // private hex to bigint
        var privateInt = new BigInteger(privateKeyHex, 16);
        // secp256k1 spec curve
        ECDomainParameters curve = KeyUtils.Defaults.DomainParameters;
        var privateKey = new ECPrivateKeyParameters(privateInt, curve);
        return Sign(message, privateKey);
    }

    public MessageSignatureHex Sign(string message, ECPrivateKeyParameters privateKey) {
        // init signer
        var signer = new ECDsaSigner(new HMacDsaKCalculator(new Sha256Digest()));
        signer.Init(true, privateKey);
        // message string to keccak256 hash
        byte[] messageHash = message.HashKeccak256();
        // sign

        BigInteger curveOrder = privateKey.Parameters.N;
        BigInteger halfCurveOrder = curveOrder.ShiftRight(1);

        while (true)
        {
            BigInteger[] signature = signer.GenerateSignature(messageHash);
            BigInteger r = signature[0];
            BigInteger s = signature[1];

            if (r.CompareTo(halfCurveOrder) <= 0)
            {
                // Ensure S is the lower of S and -S
                if (s.CompareTo(halfCurveOrder) > 0)
                {
                    s = curveOrder.Subtract(s);
                }

                BigInteger[] actualSignature = { r, s };
                return new MessageSignatureInt(actualSignature).ToHex();
            }
        }

//        BigInteger[] signature = signer.GenerateSignature(messageHash);
//        return new MessageSignatureInt(signature).ToHex();
    }

/*    public MessageSignatureHex Sign(string message, ECPrivateKeyParameters privateKey) {
        // init signer
        var signer = new ECDsaSigner(new HMacDsaKCalculator(new Sha256Digest()));
        signer.Init(true, privateKey);
        // message string to keccak256 hash
        byte[] messageHash = message.HashKeccak256();
        // sign
        BigInteger[] signature = signer.GenerateSignature(messageHash);
        return new MessageSignatureInt(signature).ToHex();
    }*/

/*    public MessageSignatureHex Sign(string message, ECPrivateKeyParameters privateKey) {
    var signer = new ECDsaSigner(new HMacDsaKCalculator(new Sha256Digest()));
    signer.Init(true, privateKey);
    
    byte[] messageHash = message.HashKeccak256();
    Console.WriteLine($"Message Hash: {BitConverter.ToString(messageHash).Replace("-", "")}");
    
    BigInteger[] signature = signer.GenerateSignature(messageHash);
    Console.WriteLine($"R: {signature[0].ToString(16)}");
    Console.WriteLine($"S: {signature[1].ToString(16)}");
    
    // Ensure low S
    if (signature[1].CompareTo(privateKey.Parameters.N.ShiftRight(1)) > 0)
    {
        signature[1] = privateKey.Parameters.N.Subtract(signature[1]);
        Console.WriteLine($"Adjusted S: {signature[1].ToString(16)}");
    }
    
    var result = new MessageSignatureInt(signature).ToHex();
    Console.WriteLine($"Final Signature: {result}");
    
    // Immediate verification
    //var verifier = new ECDsaSigner();
    //verifier.Init(false, privateKey.Parameters.G.Multiply(privateKey.D));
    //bool verified = verifier.VerifySignature(messageHash, signature[0], signature[1]);
    //Console.WriteLine($"Immediate Verification: {verified}");
    
    return result;
}*/


    public bool VerifySignature(SignedMessage signedMessage) {
        var publicHex = GetKeys()?.PublicKey
            ?? throw new KeyPairNotFoundException();
        var withKey = signedMessage.WithKey(publicHex);
        return VerifySignature(withKey);
    }

    public bool VerifySignature(SignedMessageWithKey signedMessage) {
        return VerifySignature(signedMessage.ToEC());
    }

    public bool VerifySignature(SignedMessageWithECKey signedMessage) {
        // signature hex to bigint
        var signatureInt = signedMessage.Signature.ToInt();
        // message string to keccak256 hash
        byte[] messageHash = signedMessage.Message.HashKeccak256();
        // verify
        var verifier = new ECDsaSigner();
        verifier.Init(false, signedMessage.PublicKey);
        return verifier.VerifySignature(messageHash, signatureInt.R, signatureInt.S);
    }

    public bool Associate(params SignedMessage[] messages) {
        if (messages.Length < 2) {
            throw new ArgumentException($"{nameof(messages)} length must be greater or equal to 2");
        }
        return messages.All(VerifySignature);
    }
}
